pkg/reconciler/ingress/config/istio.go

```diff
@@ -89,17 +116,120 @@ type Istio struct {
 	LocalGateways []Gateway
 }
 
-func parseGateways(configMap *corev1.ConfigMap, prefix string) ([]Gateway, error) {
+func (i Istio) Validate() error {
+	for _, gtw := range i.IngressGateways {
+		if err := gtw.Validate(); err != nil {
+			return fmt.Errorf("invalid gateway: %w", err)
+		}
+	}
+
+	for _, gtw := range i.LocalGateways {
+		if err := gtw.Validate(); err != nil {
+			return fmt.Errorf("invalid local gateway: %w", err)
+		}
+	}
+
+	return nil
+}
+
+// NewIstioFromConfigMap creates an Istio config from the supplied ConfigMap
+func NewIstioFromConfigMap(configMap *corev1.ConfigMap) (*Istio, error) {
+	ret, err := parseNewFormat(configMap)
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse configmap: %w", err)
+	}
+
+	// If the new format is not defined, try to parse the old one
+	if len(ret.IngressGateways) == 0 && len(ret.LocalGateways) == 0 {
+		ret = parseOldFormat(configMap)
+	}
+
+	err = ret.Validate()
+	if err != nil {
+		return nil, fmt.Errorf("invalid configmap: %w", err)
+	}
+
+	return ret, nil
+}
+
+func parseNewFormat(configMap *corev1.ConfigMap) (*Istio, error) {
+	ret := &Istio{}
+
+	gatewaysStr, hasGateway := configMap.Data[gatewayKey]
+	localGatewaysStr, hasLocalGateway := configMap.Data[localGatewayKey]
+
+	// New format is partially defined
+	if (!hasGateway && hasLocalGateway) || (hasGateway && !hasLocalGateway) {
+		return nil, fmt.Errorf("config is not fully defined")
+	}
+
+	// New format is not defined
+	if !hasGateway && !hasLocalGateway {
+		return ret, nil
+	}
+
+	gateways, err := parseNewFormatGateways(gatewaysStr)
+	if err != nil {
+		return ret, fmt.Errorf("failed to parse gateways")
+	}
+
+	ret.IngressGateways = gateways
+
+	localGateways, err := parseNewFormatGateways(localGatewaysStr)
+	if err != nil {
+		return ret, fmt.Errorf("failed to parse local gateways")
+	}
+
+	if len(localGateways) > 1 {
+		return ret, fmt.Errorf("only one local gateways can be defined")
```

Comment:
Maybe make the error message more clear my mentioning the key that has the problematic value